src/Compilers/CSharp/Portable/Lowering/Instrumentation/ModuleCancellationInstrumenter.cs

```diff
@@ -0,0 +1,178 @@
+        ref MethodSymbol method,
+        ref BoundExpression? receiver,
+        ref ImmutableArray<BoundExpression> arguments,
+        ref ImmutableArray<RefKind> argumentRefKindsOpt,
+        bool invokedAsExtensionMethod,
+        SimpleNameSyntax? nameSyntax)
+    {
+        Previous.InterceptCallAndAdjustArguments(ref method, ref receiver, ref arguments, ref argumentRefKindsOpt, invokedAsExtensionMethod, nameSyntax);
+
+        if (arguments is [.., { Type: { } lastArgumentType } lastArgument] &&
+            (argumentRefKindsOpt.IsDefault || argumentRefKindsOpt is [.., RefKind.None]) &&
+            lastArgumentType.Equals(throwMethod.ContainingType, TypeCompareKind.ConsiderEverything))
+        {
+            // The last argument is a CancellationToken. Replace it with the module-level token.
+            // Keep the previous expression so that side-effects are preserved.
+            arguments = [.. arguments[0..^1], factory.Sequence([lastArgument], factory.ModuleCancellationToken())];
+        }
+        else if (FindOverloadWithCancellationToken(method) is { } cancellableOverload)
+        {
+            // The method being invoked does not have a CancellationToken as the last parameter, but there is an overload that does.
+            // Invoke the other overload instead and pass in module-level token. 
+            method = cancellableOverload;
+            arguments = [.. arguments, factory.ModuleCancellationToken()];
+            argumentRefKindsOpt = argumentRefKindsOpt.IsDefault ? default : [.. argumentRefKindsOpt, RefKind.None];
+        }
+    }
+
+    /// <summary>
+    /// Find an overload whose last parameter is <see cref="CancellationToken"/> and
+    /// the parameter types and ref kinds of all other parameters match those of <paramref name="method"/>.
+    /// </summary>
+    private MethodSymbol? FindOverloadWithCancellationToken(MethodSymbol method)
+    {
+        // It's unlikely that real-world APIs have overloads that differ in dynamic,
+        // but if they do avoid selecting them since their intended use might differ.
+        var typeComparisonKind = TypeCompareKind.CLRSignatureCompareOptions & ~TypeCompareKind.IgnoreDynamic;
+
+        foreach (var member in method.ContainingType.GetMembers(method.Name))
+        {
+            if (member is MethodSymbol { Parameters: [.., { RefKind: RefKind.None, Type: { } lastParamType }] parametersWithCancellationToken } overload &&
+                overload.Arity == method.Arity &&
+                method.Parameters.Length == parametersWithCancellationToken.Length - 1 &&
+                lastParamType.Equals(throwMethod.ContainingType, TypeCompareKind.ConsiderEverything) &&
+                MemberSignatureComparer.HaveSameParameterTypes(
+                    method.Parameters.AsSpan(),
+                    typeMap1: null,
+                    parametersWithCancellationToken.AsSpan(0, method.Parameters.Length),
+                    method.TypeSubstitution,
```

Comment:
>method.TypeSubstitution [](http://example.com/codeflow?start=20&length=23)

This looks suspicious, ```method``'s type map is unlikely to substitute type parameters of the ```overload```. Are we testing scenarios with generic methods?

Reply:
Not sure what you mean. 

We have generic tests - anything missing?
https://github.com/dotnet/roslyn/pull/71896/files#diff-c2672db42ee61f37aac5cf49a22dfddaaf7d9e07d6983a9942c1c5d9a225d207R1671-R1868

Reply:
> Not sure what you mean.
> 
> We have generic tests - anything missing? 

What specific test you think cover this scenario?

Reply:
ArgumentReplacement_Overload_Generic?

Reply:
I see, `ArgumentReplacement_Overload_Generic_MatchingConstraint` test did not use the generic parameter in signature. It fails if it does.